### PR TITLE
On policy distillation

### PR DESCRIPTION
**Pull request by faresobeid**
Can use by setting this in an rl.toml config
```
teacher_gpu_ids = ...

[teacher_inference.model]
name = ...

[trainer.loss]
adv_tau = ...
teacher_tau = ...
```

Can also use an existing teacher inference instant with 
```
[orchestrator.teacher_model.client]
base_url = ["http://localhost:8001/v1"]

[orchestrator.teacher_model.model]
name = ...
```

---

> [!NOTE]
> Adds end-to-end support for on-policy distillation and an option to skip verification for pure distillation.
> 
> - **Teacher model pipeline**: New `orchestrator.teacher_model` config; RL runner can auto-start a teacher inference server via `teacher_gpu_ids`/`teacher_inference`; orchestrator computes `teacher_logprobs` via prefill and attaches to `TrainingSample`/`MicroBatch`
> - **Loss integration**: Trainer loss now supports `teacher_logprobs` with `trainer.loss.teacher_tau` and `trainer.loss.adv_tau`; logs `teacher_kl`; minor fixes to importance ratio/sequence handling
> - **Buffer control**: `orchestrator.buffer.skip_verification` disables rubric scoring and related features (rewards set to 0)
> - **Batch/transport updates**: Carry `temperature` explicitly and include optional `teacher_logprobs` through packing/padding to GPU tensors
> - **Docs and validation**: New `docs/on_policy_distillation.md`; config validators to auto-configure teacher server/client and prevent invalid combos; tests updated
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 9b6383cb88162fa5e9c616d32ee1d20e86d5b7f7. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by willccbb** [comment]
Verifiers PR for skipping rubric scoring: https://github.com/PrimeIntellect-ai/verifiers/pull/645

File: src/prime_rl/orchestrator/config.py [hunk 1]
```diff
@@ -445,6 +445,17 @@ class BufferConfig(BaseConfig):
...
+    score_rollouts: Annotated[
+        bool,
+        Field(
+            description=(
+                "Whether to score rollouts using the environment's rubric. "
+                "If False, rewards are always set to 0, online_difficulty_filtering is disabled, "
+                "and easy/hard thresholds are not used."
+            ),
+        ),
+    ] = True
```
Comment (samsja) [inline src/prime_rl/orchestrator/config.py]: would inverse the logic, should be something like "skip_verification"

File: src/prime_rl/trainer/rl/loss.py [hunk 2]
```diff
@@ -112,6 +115,7 @@ def compute_loss(
 
         importance_ratio = seq_importance_ratio if loss_config.ratio_type == "sequence" else token_importance_ratio
 
+        advantages = loss_config.adv_tau * advantages + loss_config.teacher_tau * teacher_kl.detach()
```
Comment (samsja) [inline src/prime_rl/trainer/rl/loss.py]: hmmm its not clear to me why we need to have the teacher_kl here.

Why not putting this into the advantages calculation on the orch side ?

Comment (faresobeid) [reply src/prime_rl/trainer/rl/loss.py]: bcs we use the trainer_logprobs. Could use inference_logprobs but no reason to, trainer_logprobs should prob be a bit better, and here we can choose which one to use easily

Comment (samsja) [reply src/prime_rl/trainer/rl/loss.py]: makes sense

**Comment by faresobeid** [comment]
image
Purple is with a bit of teacher_kl. reverse-text is a special env as rewards are already non-binary so already dense to some extent

File: src/prime_rl/orchestrator/config.py [hunk 3]
```diff
@@ -457,6 +468,15 @@ def validate_env_ratios(self):
...
         return self
 
+    @model_validator(mode="after")
+    def auto_setup_skip_verification(self):
+        """When scoring is disabled, auto-disable features that depend on rewards."""
+        if self.skip_verification:
+            self.online_difficulty_filtering = False
+            self.easy_threshold = None
+            self.hard_threshold = None
+        return self
```
Comment (samsja) [inline src/prime_rl/orchestrator/config.py:490]: I would rather failed here that changing the field

File: src/prime_rl/orchestrator/orchestrator.py [hunk 4]
```diff
@@ -320,6 +335,21 @@ async def orchestrate(config: OrchestratorConfig):
             f"Converted {len(train_rollouts)} training rollouts to {len(train_examples)} training examples using {config.trajectory_strategy} strategy"
         )
 
+        # Compute teacher logprobs if teacher model is configured
+        teacher_logprobs_time = 0
+        if teacher_clients is not None and teacher_model_name is not None:
```
Comment (samsja) [inline src/prime_rl/orchestrator/orchestrator.py]:
```suggestion
        if config.teacher_model is not None:
```

File: src/prime_rl/trainer/rl/data.py [hunk 5]
```diff
@@ -21,6 +21,7 @@ class TensorMicroBatch(TypedDict):
     position_ids: Int[Tensor, "batch seq"]
     advantages: Float[Tensor, "batch seq"]
     inference_logprobs: Float[Tensor, "batch seq"]
+    teacher_logprobs: Float[Tensor, "batch seq"]
```
Comment (samsja) [inline src/prime_rl/trainer/rl/data.py]: would rather make this one None instead of 0 by default

**Review by samsja — APPROVED** [review]
lgtm great work
